### PR TITLE
[RLlib] RLModule: Add `TargetNetworkAPI` and implement for APPO and SAC.

### PR DESCRIPTION
**Pull request by sven1977**
RLModule: Add`TargetNetworkAPI` and implement for APPO and SAC.

Our goal for custom, user-written RLModules is that users should only have to write "a simple torch.nn.Module", subclassing from `TorchRLModule` (defining the three different forward passes). No subclassing from complex algo-specific RLModules should be required, unless(!) users want to use our RLlib default models and develop on top of these.
Our current algo-specific RLModule classes are too complex and scare users away from subclassing and implementing them. Also, being able to switch more easily between algos, once a custom network architecture has been written out, is currently extremely difficult.

We instead propose simple APIs (if possible w/o any code/implementations, just abstract method definitions), e.g. this `TargetNetworkAPI` here, to be plugged into the users' custom RLModules. This way, a algo-specific Learner can communicate to the user that the provided custom RLModule lacks implementing API xyz and all the user then has to do is inherit from API xyz as well and implement a couple of abstract methods described by this API.

TODO: Translate DQN into this new API.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by simonsays1980 — APPROVED** [review]
LGTM. There are still some `requires_grad` that - I think - could improve performance. Added another comment in regard to updating target networks.

File: rllib/algorithms/sac/torch/sac_torch_rl_module.py [hunk 1]
```diff
@@ -29,51 +30,9 @@ def setup(self):
...
-            # We do not want to train the target networks. Instead, we sync them
-            # with the actual (trained) ones.
-            self.qf_target_encoder.requires_grad_(False)
-            self.qf_target_encoder.load_state_dict(self.qf_encoder.state_dict())
-            self.qf_target.requires_grad_(False)
-            self.qf_target.load_state_dict(self.qf.state_dict())
-
-            # If necessary, also synchronize the twin networks.
-            if self.twin_q:
-                self.qf_target_twin_encoder.requires_grad_(False)
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/sac/torch/sac_torch_rl_module.py:40]: This step is already done in the `make_target_network()` utility, though. Same as the weight-sync.

File: rllib/core/rl_module/torch/torch_rl_module.py [hunk 2]
```diff
@@ -178,6 +178,8 @@ def unwrapped(self) -> "RLModule":
         return self.module
 
 
+# TODO (sven): Deprecate in favor of TargetNetworkAPI (only DQN still using
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/core/rl_module/torch/torch_rl_module.py:181]: Good question :) 
DQN was to complex to put it into the PR as well. I'll need to do another follow-up PR to include it. No particular other reason.

File: rllib/core/rl_module/apis/target_network_api.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import abc
+from typing import Any, Dict, List, Tuple
+
+from ray.rllib.utils.typing import NetworkType
+
+
+class TargetNetworkAPI(abc.ABC):
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/core/rl_module/apis/target_network_api.py:7]: Correct: WIP. DQN is missing. I didn't want to include it as well and see, whether the API makes sense first (with SAC and APPO).

File: rllib/algorithms/dqn/torch/dqn_rainbow_torch_rl_module.py [hunk 4]
```diff
@@ -35,13 +35,10 @@ def setup(self):
         # space is a flat space we can use a noisy encoder.
         self.uses_noisy_encoder = isinstance(self.encoder, TorchNoisyMLPEncoder)
 
-        # If we have target networks ,we need to make them not trainable.
+        # If not an inference-only module (e.g., for evaluation), set up the
+        # parameter names to be removed or renamed when syncing from the state dict
+        # when syncing.
         if not self.config.inference_only:
-            self.target_encoder.requires_grad_(False)
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dqn/torch/dqn_rainbow_torch_rl_module.py:40]: see my comment below. We still do this, but in `make_target_network` (along with the initial weight-sync).

**Pull request merged**